### DISCUSSION TITLE
Triggering multiple recorder blocks at different times based on IO channels.

### DISCUSSION THREAD

**Discussion by magnus-haw** [LaboratoireMecaniqueLille/crappy#31] [Q&A] [answered]
Hi, this is a cool project! I'm new to crappy and have an experiment where I want to start recording data from two separate digitizers at different times based on specific data channels. Is this possible or do all recording blocks need to run for the whole experiment?

**Comment by WeisLeDocto** [comment]
Hi, thank you for your interest in Crappy ! It's always nice to find out we have users in various locations.

Currently, the Recorder Blocks start acquiring data as soon as the experiment begins and only stop when the experiment ends.
They only record what they're being sent though, so if no data is acquired or sent then no data is saved.

May I ask you why you want to start and stop recording data during the experiment ?
Is it not possible to continuously acquire data from both digitizers and consider only the relevant parts when post-processing ?

Weis

**Reply by magnus-haw** [reply]
It is possible to acquire data from all sources for the whole experiment, however, it would be nice to not have to delete >99% of our data each time. We have a system with two time scales- we need monitoring diagnostics for the system to reach a steady state (30 min, 50Hz DAQ) and then high-speed diagnostics for fast insertion measurements (0.1 sec, 8 kHz). We're currently using a LabJack T7 (streaming @8 kHz) for the fast insertion data. Currently we operate these systems separately but it would be nice to be able to control both via crappy. 

Is there a way for the Recorder blocks to save data only when a certain condition is satisfied (e.g., while "myvar > 1")? If not, can I easily write a custom block for this? Perhaps a custom modifier that only passes data when a condition is met?

**Comment by Dad0u** [comment, marked as answer]
Hi @magnus-haw !
Recording data on a specific condition seems to be the perfect job for [modifiers](https://crappy.readthedocs.io/en/stable/tutorials/custom_blocks.html#e-modifiers). You can use a modifier to alter or drop the data going through a link.
For the example you described (only recording when a given value is above a threshold) you can use a simple function as a modifier like this :
```
def only_above_0(data):
    """
    Only returns the data if the value of the label src is > 0
    """
    if data['label_to_compare'] > 0:
        return data
    return None
```

Then, the link between your data source and Recorder block can be done with `crappy.link(data_source, recorder_block, modifier=only_above_0)`

Hope this helps !
Victor

**Reply by jeffwitz** [reply]
To complete Victor answer, I think that there is an example that uses a personalized modifier [here](https://github.com/LaboratoireMecaniqueLille/crappy/blob/master/Examples/stream_t7.py). Using that as a starting point, you can create a function that pick values (or data chunks means) for the first 30mins and then detect you insertion. But if you are using two different devices for this two different phases, then I think that you have to create two different functions for your two different devices used as modifiers that can be send to the same graph and recorder.
Be cautious if you want to use the hdf recorder, if something goes wrong the .h5 file can be corrupted, and there is no way (easy or hard) to recover data ...
